pkg/kubelet/secret/secret_manager.go

```diff
@@ -155,3 +160,92 @@ func NewWatchingSecretManager(kubeClients []clientset.Interface) Manager {
 		manager: manager.NewWatchBasedManager(listSecret, watchSecret, newSecret, gr, getSecretNames),
 	}
 }
+
+type byHostSecretManager struct {
+	kubeClients []clientset.Interface
+	hostName    string
+	stores      []cache.Store
+}
+
+// ensure this is the same as cache.MetaNamespaceKeyFunc
+func (s *byHostSecretManager) key(tenant, namespace, name string) string {
+	result := name
+	if len(namespace) > 0 {
+		result = namespace + "/" + result
+	} else {
+		result = metav1.NamespaceDefault + "/" + result
+	}
+	if len(tenant) > 0 && tenant != metav1.TenantSystem {
+		result = tenant + "/" + result
+	} else {
+		result = metav1.TenantSystem + "/" + result
+	}
+	return result
+}
+
+func (s *byHostSecretManager) GetSecret(tenant, namespace, name string) (*v1.Secret, error) {
+	key := s.key(tenant, namespace, name)
+	klog.Infof("get secret: %s", key)
+	for _, store := range s.stores {
+		klog.Infof("debug: dump store keys: [%v]", store.ListKeys())
+		object, _, err := store.GetByKey(key)
+		if err != nil {
+			return nil, err
+		}
+		if object, ok := object.(*v1.Secret); ok {
+			return object, nil
+		}
+		return nil, fmt.Errorf("unexpected object type: %v", object)
+
+	}
+	return nil, fmt.Errorf("secret not found: %s-%s-%s", tenant, namespace, name)
+}
+
+func (s *byHostSecretManager) RegisterPod(pod *v1.Pod) {
+}
+
+func (s *byHostSecretManager) UnregisterPod(pod *v1.Pod) {
+}
+
+func MetaNamespaceKeyFunc(obj interface{}) (string, error) {
+	meta, err := meta.Accessor(obj)
+	if err != nil {
+		return "", fmt.Errorf("object has no meta: %v", err)
+	}
+
+	metaKey := meta.GetName()
+	if len(meta.GetNamespace()) > 0 {
+		metaKey = meta.GetNamespace() + "/" + metaKey
+	} else {
+		metaKey = metav1.NamespaceDefault + "/" + metaKey
+	}
+
+	if len(meta.GetTenant()) > 0 {
+		metaKey = meta.GetTenant() + "/" + metaKey
+	} else {
+		metaKey = metav1.TenantSystem + "/" + metaKey
+	}
+
+	return metaKey, nil
+}
+
+func NewByHostWatchingSecretManager(kubeClients []clientset.Interface, hostName string) Manager {
+
+	klog.Infof("create secret manager for host: %s", hostName)
+	stores := make([]cache.Store, len(kubeClients))
+
+	for i, tenantPartitionClient := range kubeClients {
```

Comment:
This push all secrets tied to namespaces to all node, could really be a security concern.

Reply:
Also greatly increased secret data population.

Reply:
i agree. as it is commented, the list/watch will be node based once the secret controller logic is in